x/mongo/driver/operation/hello.go

```diff
@@ -113,6 +122,417 @@ func (h *Hello) Result(addr address.Address) description.Server {
+		region := os.Getenv(envVar)
+		if region == "" {
+			return nil
+		}
+
+		return func(dst []byte) []byte {
+			key := "region"
+
+			return bsoncore.AppendStringElement(dst, key, region)
+		}
+	}
+
+	addTimeout := func(envVar string) func([]byte) []byte {
+		timeout := os.Getenv(envVar)
+		if timeout == "" {
+			return nil
+		}
+
+		timeoutInt64, err := strconv.ParseInt(timeout, 10, 32)
+		if err != nil {
+			return nil
+		}
+
+		timeoutInt32 := int32(timeoutInt64)
+		return func(dst []byte) []byte {
+			key := "timeout_sec"
+
+			return bsoncore.AppendInt32Element(dst, key, timeoutInt32)
+		}
+	}
+
+	addURL := func(envVar string) func([]byte) []byte {
+		url := os.Getenv(envVar)
+		if url == "" {
+			return nil
+		}
+
+		return func(dst []byte) []byte {
+			return bsoncore.AppendStringElement(dst, "url", url)
+		}
+	}
+
+	fns := []func([]byte) []byte{
+		func(dst []byte) []byte {
+			return bsoncore.AppendStringElement(dst, "name", name)
+		},
+	}
+
+	switch name {
+	case envNameAWSLambda:
+		fns = append(fns, addMem(envVarAWSLambdaFunctionMemorySize))
+		fns = append(fns, addRegion(envVarAWSRegion))
+	case envNameGCPFunc:
+		fns = append(fns, addMem(envVarFunctionMemoryMB))
+		fns = append(fns, addRegion(envVarFunctionRegion))
+		fns = append(fns, addTimeout(envVarFunctionTimeoutSec))
+	case envNameVercel:
+		fns = append(fns, addRegion(envVarVercelRegion))
+		fns = append(fns, addURL(envVarVercelURL))
+	}
+
+	const envKey = "env"
+	return appendSubDocument(dst, maxLen, envKey, fns)
+}
+
+// appendClientOS appends the OS metadata to dst. If the document exceeds the
+// maximum length, the OS is omitted.
+func appendClientOS(dst []byte, maxLen int) ([]byte, error) {
+	if dst == nil || maxLen <= 0 {
+		return dst, nil
+	}
+
+	const osKey = "os"
+	var err error
+
+	dst, err = appendSubDocument(dst, maxLen, osKey, []func([]byte) []byte{
+		func(dst []byte) []byte {
+			const key = "type"
+
+			return bsoncore.AppendStringElement(dst, key, runtime.GOOS)
+		},
+		func(dst []byte) []byte {
+			const key = "architecture"
+
+			return bsoncore.AppendStringElement(dst, key, runtime.GOARCH)
+		},
+	})
+
+	return dst, err
+}
+
+// appendClientPlatform appends the platform metadata to dst. If the document
+// exceeds the maximum length, the platform is omitted.
+func appendClientPlatform(dst []byte, maxLen int) []byte {
+	if dst == nil || maxLen <= 0 {
+		return dst
+	}
+
+	const platformKey = "platform"
+	return appendWithMaxLen(dst, maxLen, func(dst []byte) []byte {
+		return bsoncore.AppendStringElement(dst, platformKey, runtime.Version())
+	})
+}
+
+// encodeClientMetadata encodes the client metadata into a BSON document. maxLen
+// is the maximum length the document can be. If the document exceeds maxLen,
+// then an empty byte slice is returned. If there is not enough space to encode
+// a document, the document is truncated and returned.
+//
+// This function attempts to build the following document, prioritizing upto the
+// givien order:
```

Comment:
The specification recommends:

> Drivers MUST validate these values and truncate or omit driver provided values if necessary. Implementors SHOULD cumulatively update fields in the following order until the document is under the size limit:
> 1. Omit fields from env except env.name.
> 2. Omit fields from os except os.type.
> 3. Omit the env document entirely.
> 4. Truncate platform.

I suggest a different approach. Do not pass `maxLen` to the `appendClient{AppName,Driver,OS,Platform,Env}` functions. Instead, try to append everything to `client`. If the final `client` document exceeds `maxLen`, retry following the suggested omisions.

[This branch](https://github.com/kevinAlbs/mongo-go-driver/tree/retry_append) contains an implementation of the idea. Here is the most relevant part:

```go
func encodeClientMetadata(appname string, maxLen int) ([]byte, error) {
	dst := make([]byte, 0, maxLen)

	omitEnvFields := false
	omitOSFields := false
	omitEnvDocument := false
	truncatePlatform := false

retry:

	idx, dst := bsoncore.AppendDocumentStart(dst)

	var err error
	dst, err = appendClientAppName(dst, appname)
	if err != nil {
		return dst, err
	}

	dst, err = appendClientDriver(dst)
	if err != nil {
		return dst, err
	}

	dst, err = appendClientOS(dst, omitOSFields)
	if err != nil {
		return dst, err
	}

	if !truncatePlatform {
		dst = appendClientPlatform(dst)
	}

	if !omitEnvDocument {
		dst, err = appendClientEnv(dst, omitEnvFields)
		if err != nil {
			return dst, err
		}
	}

	dst, err = bsoncore.AppendDocumentEnd(dst, idx)
	if err != nil {
		return dst, err
	}

	if len(dst) > maxLen {
		// The final "dst" slice is greater than the maximum length. Retry following guidance from the Handshake specification:
		// """
		//    Implementors SHOULD cumulatively update fields in
		//    the following order until the document is under the size limit
		//    1. Omit fields from ``env`` except ``env.name``.
		//    2. Omit fields from ``os`` except ``os.type``.
		//    3. Omit the ``env`` document entirely.
		//    4. Truncate ``platform``.
		// """
		dst = dst[:0]

		if !omitEnvFields {
			omitEnvFields = true
			goto retry
		}

		if !omitOSFields {
			omitOSFields = true
			goto retry
		}

		if !omitEnvDocument {
			omitEnvDocument = true
			goto retry
		}

		if !truncatePlatform {
			truncatePlatform = true
			goto retry
		}

		// There is nothing left to update.
		// Return an empty slice to tell caller not to append a `client` document.
		return dst[:0], nil
	}

	return dst, nil
}
```